src/NHibernate/Loader/Loader.cs

```diff
@@ -1708,60 +1714,54 @@ private CacheableResultTransformer CreateCacheableResultTransformer(QueryParamet
 				queryParameters.HasAutoDiscoverScalarTypes, SqlString);
 		}
 
-		internal IList GetResultFromQueryCache(ISessionImplementor session, QueryParameters queryParameters,
-											  ISet<string> querySpaces, IQueryCache queryCache,
-											  QueryKey key)
+		internal bool CanGetFromCache(ISessionImplementor session, QueryParameters queryParameters)
```

Comment:
This method doesn't handle per query `SetCacheMode` calls. Which actually now is used in tests (as no op but might lead to confusion that it actually can be used in Future):

https://github.com/nhibernate/nhibernate-core/blob/7ce79eda3a184e3225fcf6fa4ed3c15fc27dc0ab/src/NHibernate.Test/Futures/QueryBatchFixture.cs#L354-L355

Maybe this query cache mode needs to be moved to `QueryParameters` and handled similarly as `QueryParameters.ReadOnly` flag?

So we should move this code to `QueryParameters.CanGetFromCache(session)` method. And for parity introduce `QueryParameters.CanPutToCache(session)`?

Reply:
Indeed, the queries individual cache mode has never been handled for future and multi-query/criteria, but we could. I do not think it should be done here, but better as another PR.

Reply:
I think it will not work, so better stick to YAGNI, not doing this in advance.

`CacheMode` needs to be temporarily switched on the session instead, because many cache interactions triggered deeply in the loading implementation have no access to query parameters.

Reply:
For this PR I'm just saying that it's better to have consistent way for readonly check for query and cache mode check for query. So from this point - what will not work?
We use `queryParameters.IsReadOnly(session)` for readonly check. So why introduce new logically similair method in `Loader` instead? It can be still marked internal here if you think this "cache mode per query" won't work.

Reply:
These cases are not comparable in my opinion:
- `IsReadonly` is a bit messy, originating from times where nullable were not avaliable. It is also there to hide the complexity of replacing nullability by an additional boolean.
- `CacheMode` is not a boolean but a flag enumeration value. It conveys more state than a boolean. `CanGetFromCache` is just a small part of it, because by example it tells nothing about what should be done with entities assembled from their id cached in query result, whether they should themselves be put into the cache or not if they were missing in it when re-assembled.
  So it would be just a partial part of the information we actually need to handle properly cache mode, and we will have to put more somehow somewhere. It then looks likely to me that `CanGetFromCache` will just be redundant and not needed.

Reply:
I'm not talking about how data needs to be stored. But rather how it should be used in code.
Calling `queryParameters.IsReadOnly(session)` for checking query read-only flag looks pretty clean to me. How messy its implementation is irrelevant.
And as we need now method  `CanGetFromCache` I think place and synatx should be the same. But let's not spend more time on it. If you are not convinced - let's keep it as is for time being.

Reply:
You were right with the need for `CanGetFromCache` and `CanPutToCache`, but I still think it was better to do it in a dedicated PR.